pkg/trie/triedb/proof/generate.go

```diff
@@ -0,0 +1,455 @@
+		var key = []byte(keys[i])
+		var keyNibbles = nibbles.KeyLEToNibbles(key)
+
+		err := unwindStack(stack, proofNodes, &keyNibbles)
+		if err != nil {
+			return nil, err
+		}
+
+		// Traverse the trie recording the visited nodes
+		recorder := triedb.NewRecorder()
+		trie := triedb.NewTrieDB(rootHash, db, triedb.WithRecorder(recorder))
+		trie.SetVersion(trieVersion)
+		trie.Get(key)
+
+		recordedNodes := triedb.NewRecordedNodesIterator(recorder.Drain())
+
+		// Skip over recorded nodes already on the stack.
+		for i := 0; i < stack.Len(); i++ {
+			nextEntry := stack.At(i)
+			nextRecord := recordedNodes.Peek()
+
+			if nextRecord == nil || !bytes.Equal(*nextEntry.nodeHash, nextRecord.Hash) {
+				break
+			}
+
+			recordedNodes.Next()
+		}
+
+		// Descend in trie collecting nodes until find the value or the end of the path
+	loop:
+		for {
+			var nextStep step
+			var entry *stackEntry
+			if stack.Len() > 0 {
+				entry = stack.Back()
+			}
+			if entry == nil {
+				nextStep = stepDescend{childPrefixLen: 0, child: nodeHandleHash{hash: rootHash.ToBytes()}}
+			} else {
+				var err error
+				nextStep, err = matchKeyToNode(
+					entry.node,
+					&entry.omitValue,
+					&entry.childIndex,
+					keyNibbles,
+					len(entry.prefix),
+					recordedNodes,
+				)
+
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			switch s := nextStep.(type) {
+			case stepDescend:
+				childPrefix := keyNibbles[:s.childPrefixLen]
+				var childEntry *stackEntry
+				switch child := s.child.(type) {
+				case nodeHandleHash:
+					childRecord := recordedNodes.Next()
+
+					if !bytes.Equal(childRecord.Hash, child.hash) {
+						panic("hash mismatch")
+					}
+
+					outputIndex := len(proofNodes)
+
+					// Insert a placeholder into output which will be replaced when this
+					// new entry is popped from the stack.
+					proofNodes = append(proofNodes, []byte{})
+					childEntry, err = newStackEntry(
+						childPrefix,
+						childRecord.Data,
+						&childRecord.Hash,
+						&outputIndex,
+					)
+
+					if err != nil {
+						return nil, err
+					}
+				case nodeHandleInline:
+					if len(child.data) > common.HashLength {
+						return nil, errors.New("invalid hash length")
+					}
+					childEntry, err = newStackEntry(
+						childPrefix,
+						child.data,
+						nil,
+						nil,
+					)
+					if err != nil {
+						return nil, err
+					}
+				}
+				stack.PushBack(childEntry)
+			default:
+				recordedNodes.Next()
+				break loop
+			}
+		}
+	}
+
+	err = unwindStack(stack, proofNodes, nil)
+	if err != nil {
+		return nil, err
+	}
+	return proofNodes, nil
+}
+
+// / Unwind the stack until the given key is prefixed by the entry at the top of the stack. If the
+// / key is nil, unwind the stack completely. As entries are popped from the stack, they are
+// / encoded into proof nodes and added to the finalized proof.
+func unwindStack(
+	stack *deque.Deque[*stackEntry],
+	proofNodes [][]byte,
+	maybeKey *[]byte,
```

Comment:
what is maybeKey? Perhaps could use a more intuitive name?

Reply:
Because it could be 'nil' but I can rename it to just 'key'